### PR TITLE
Feat/gdoc exportv2 pod

### PR DESCRIPTION
**Pull request by Harshita-mindfire**
# Pull Request Checklist

You can go to [dendron pull requests](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html) to see full details for items in this checklist.

### Quality Assurance
- [ ] Add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [ ] Make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [ ] Do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] If your tests involve running [[manual Testing|dendron://dendron.dendron-site/dendron.dev.qa.test#manual-testing]], make sure to update [[Test Workspace|dendron://dendron.dendron-site/dendron.dev.ref.test-workspace]] with any necessary notes/additions to perform manual test ^9jtWc6ov340p
- [ ] After you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: If you running MacOS or Linux, pay special attention to the Windows output and vice versa if you are developing on Windows

#### Special Cases
- [ ] If your tests changes an existing snapshot, make sure that snapshots are [updated](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#updating-test-snapshots)
- [ ] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), make sure that it is included in [test-workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace). We use this to manually inspect new changes and for auto regression testing 

### Docs
- [ ] Make sure that the PR title follows our [commit style](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html#commit-style)
- [ ] Please summarize the feature or impact in 1-2 lines in the PR description
- [ ] If your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

Example PR Description
```markdown
# feat: capitalize all foos

This changes capitalizes all occurrences of `foo` to `Foo` 

Docs PR: <URL_TO_DOCS_PR>
```

### First Time PR
- [ ] Sign the [CLA](https://en.wikipedia.org/wiki/Contributor_License_Agreement) which will be prompted by our github bot after you submit the PR
- [ ] Add your [discord](https://discord.gg/AE3NRw9) alias in the review so that we can give you the [horticulturalist](https://wiki.dendron.so/notes/7c00d606-7b75-4d28-b563-d75f33f8e0d7.html#horticulturalist) badge in our community

### Git
- [ ] Make sure your branch names adhere to our commit style [[#^6zdCscSXs1MM]]
    - All PRs should start with `[feat|fix|enhance|]/[{description-of-pr-in-kebab-case}]`
        - `eg. feat/add-thisthing`

### Analytics
- [ ] If you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Github Issue
- [ ] If this pull request is addressing an existing issue, make sure to link this PR to the issue that it is resolving.
- [ ] If the resolution comes with a document update, link the docs PR to the issue as well.

### CODE REVIEW

File: packages/plugin-core/src/components/pods/PodControls.ts [hunk 1]
```diff
@@ -288,4 +306,46 @@ export class PodUIControls {
...
     return qp;
   }
+
+  /**
+   * Create a new configuration file for a service connection
+   * @param serviceType
+   * @param id a unique ID to identify this connection
+   * @returns full path to the newly created config file
+   */
+  public static async createNewConfig({
```
Comment (jonathanyeung) [inline packages/plugin-core/src/components/pods/PodControls.ts:316]: Why'd this function get moved to PodControls?

File: packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import path from "path";
+import * as vscode from "vscode";
+import { window } from "vscode";
+import { PodUIControls } from "../../components/pods/PodControls";
+import { VSCodeUtils } from "../../utils";
+import { getDWorkspace, getEngine, getExtension } from "../../workspace";
+import { BaseExportPodCommand } from "./BaseExportPodCommand";
+
+/**
+ * VSCode command for running the Airtable Export Pod. It is not meant to be
```
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts:25]: Update comment from saying "Airtable"

File: packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      return;
+    }
+    if (exportReturnValue.data?.documentId) {
+      await this.updateNoteWithDocumentId({
+        documentId: exportReturnValue.data?.documentId,
+        note: payload,
+      });
+    }
+    vscode.window.showInformationMessage(
+      "Finished running Google Docs export pod."
```
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts:160]: nit: can we expand this information message to give an indication of the gdoc file name that was exported to?

File: packages/pods-core/src/v2/pods/export/GoogleDocsExportPodV2.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+export type GoogleDocsExportReturnType = RespV2<{
+  /**
+   * Document Id of the exported Note
+   */
+  documentId?: string;
+}>;
+
+/**
+ * GDoc Export Pod (V2 - for compatibility with Pod V2 workflow). Supports
+ * both exportText() and exportNote()
```
Comment (jonathanyeung) [inline packages/pods-core/src/v2/pods/export/GoogleDocsExportPodV2.ts:32]: update comment - doesn't look like we support `exportText()` yet?

**Review by jonathanyeung — CHANGES_REQUESTED** [review]
I know this is still marked as Draft, but I left some early comments :)

File: packages/plugin-core/src/components/pods/PodControls.ts (hunk 1 above)
Comment (Harshita-mindfire) [reply packages/plugin-core/src/components/pods/PodControls.ts:316]: Had to discuss regarding this function. To migrate gdoc ouath flow in `ExternalConnectionManager` from `plugin-core/pods.ts`  we require google oauth client id which is in plugin core. I wasn't sure about the best way to provide client id to connection manager when the external service is google.

File: packages/pods-core/src/v2/pods/export/GoogleDocsExportPodV2.ts (hunk 4 above)
Comment (Harshita-mindfire) [reply packages/pods-core/src/v2/pods/export/GoogleDocsExportPodV2.ts:32]: Yes, will update this.

File: packages/plugin-core/src/components/pods/PodControls.ts (hunk 1 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/components/pods/PodControls.ts:316]: I don't think we need to make the client ID a part of the serialized config. The Client ID should be reused across all gdoc connections.

File: packages/plugin-core/src/commands/pods/ConfigureServiceConnection.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    this.L.info({ ctx, opts });
+    const { serviceType } = opts;
+    let configFilePath: string;
+    const mngr = new ExternalConnectionManager(getExtension().podsDir);
+    const existingConnections = await mngr.getAllConfigsByType(serviceType);
+    if (existingConnections.length > 0) {
+      const items = existingConnections.map<QuickPickItem>((value) => {
+        return { label: value.connectionId };
+      });
+      const userChoice = await window.showQuickPick(items, {
```
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/pods/ConfigureServiceConnection.ts:44]: Maybe for vNext, we can add another option of 'Create New Service Connection' into this picker

File: packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      });
+    }
+    const gdocLink = `https://docs.google.com/document/d/${exportReturnValue.data?.documentId}`;
+    clipboard.writeText(gdocLink);
+    vscode.window.showInformationMessage(
+      `Finished running Google Docs export pod. The doc ${payload.fname} is created in your selected google docs account and link is copied to the clipboard.`
+    );
+  }
+
+  async updateNoteWithDocumentId(opts: {
```
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts:192]: Does this need to be public? If not, please mark as private. 

If it's visible for testing, add a comment saying "Visible only for testing purposes"

File: packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      await this.updateNoteWithDocumentId({
+        documentId,
+        revisionId,
+        note: payload,
+      });
+    }
+    const gdocLink = `https://docs.google.com/document/d/${exportReturnValue.data?.documentId}`;
+    clipboard.writeText(gdocLink);
+    vscode.window.showInformationMessage(
+      `Finished running Google Docs export pod. The doc ${payload.fname} is created in your selected google docs account and link is copied to the clipboard.`
```
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/pods/GoogleDocsExportPodCommand.ts:188]: let's shorten it a bit:

"The doc ${payload.fname} has been created. Its link has been copied to the clipboard."

File: packages/pods-core/src/v2/ConfigFileUtils.ts [hunk 8]
```diff
@@ -67,4 +73,44 @@ export class ConfigFileUtils {
     }
     return fPath;
   }
+
+  static createExportConfig(opts: { required: string[]; properties: any }) {
```
Comment (jonathanyeung) [inline packages/pods-core/src/v2/ConfigFileUtils.ts:77]: Let's try to add some typing to this.  I think if we add a generic param T, and have something like 

`opts: { required: (keyof T)[]; properties: T}`

_Exact typing syntax needs to be fixed_

File: packages/pods-core/src/v2/ConfigFileUtils.ts [hunk 9]
```diff
@@ -67,4 +73,44 @@ export class ConfigFileUtils {
...
+        podType: {
+          description: "type of pod",
+          type: "string",
+        },
+        ...opts.properties,
+      },
+    };
+  }
+
+  static getConfigSchema(podType: PodV2Types): any {
```
Comment (jonathanyeung) [inline packages/pods-core/src/v2/ConfigFileUtils.ts:104]: Let's not return `any` - at the very least, we should be able to do `JSONSchemaType<any>`?

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

File: packages/pods-core/src/v2/ConfigFileUtils.ts (hunk 9 above)
Comment (jonathanyeung) [reply packages/pods-core/src/v2/ConfigFileUtils.ts:104]: Chatted offline to resolve this one later.

**Review by jonathanyeung — APPROVED** [review]

**Pull request merged**